### PR TITLE
[Foundation] Break out of the delegate when we have canceled a task. Fixes #9132

### LINKED ISSUE #9132: App crash due to NullReferenceException in NSUrlSessionHandlerDelegate.SetResponse
### Steps to Reproduce

Unknown. It happened to me once while debugging, and we have several crash reports in App Center that look to be the same issue, but I haven't seen a consistent reproduction. This is in an iOS app running on iOS 13.

### Expected Behavior
`NSUrlSessionHandlerDelegate.SetResponse` doesn't crash the app with a NullReferenceException
### Actual Behavior
`NSUrlSessionHandlerDelegate.SetResponse` crashes the app with a NullReferenceException. `inflight.CancellationTokenSource` was null on [this line](https://github.com/xamarin/xamarin-macios/blob/edd8a2c8963063ca0ab3357e251944f4aa053fa8/src/Foundation/NSUrlSessionHandler.cs#L684):

```C#
if (inflight.CancellationTokenSource.Token.IsCancellationRequested)
```

### Environment

```
=== Visual Studio Community 2019 for Mac ===

Version 8.6.5 (build 23)
Installation UUID: bca81352-ad6e-4861-a3e7-cec8b4172cb6
	GTK+ 2.24.23 (Raleigh theme)
	Xamarin.Mac 6.18.0.23 (d16-6 / 088c73638)

	Package version: 610000104

=== Mono Framework MDK ===

Runtime:
	Mono 6.10.0.104 (2019-12/5d03a6fe116) (64-bit)
	Package version: 610000104

=== Roslyn (Language Service) ===

3.6.0-3.20210.9+4eafdcb1bcbd8d3573f2ba6065e56d9b9ce4f8a3

=== NuGet ===

Version: 5.6.0.6591

=== .NET Core SDK ===

SDK: /usr/local/share/dotnet/sdk/3.1.301/Sdks
SDK Versions:
	3.1.301
	3.1.300
	3.1.201
	3.1.200
	3.1.102
	3.1.101
	3.1.100
	3.0.101
	3.0.100
	2.2.300
	2.2.103
	2.1.505
	2.1.500
	2.1.302
MSBuild SDKs: /Library/Frameworks/Mono.framework/Versions/6.10.0/lib/mono/msbuild/Current/bin/Sdks

=== .NET Core Runtime ===

Runtime: /usr/local/share/dotnet/dotnet
Runtime Versions:
	3.1.5
	3.1.4
	3.1.3
	3.1.2
	3.1.1
	3.1.0
	3.0.1
	3.0.0
	2.2.5
	2.2.3
	2.2.1
	2.1.19
	2.1.18
	2.1.17
	2.1.16
	2.1.15
	2.1.14
	2.1.13
	2.1.9
	2.1.6
	2.1.2

=== Xamarin.Profiler ===

Version: 1.6.12.29
Location: /Applications/Xamarin Profiler.app/Contents/MacOS/Xamarin Profiler

=== Updater ===

Version: 11

=== Xamarin.Android ===

Version: 10.3.1.4 (Visual Studio Community)
Commit: xamarin-android/d16-6/3a10de9
Android SDK: /Users/stefanglr/Library/Developer/Xamarin/android-sdk-macosx
	Supported Android versions:
		6.0 (API level 23)
		7.1 (API level 25)
		8.1 (API level 27)

SDK Tools Version: 26.1.1
SDK Platform Tools Version: 29.0.5
SDK Build Tools Version: 28.0.3

Build Information: 
Mono: 165f4b0
Java.Interop: xamarin/java.interop/d16-6@2cab35c
ProGuard: xamarin/proguard/master@905836d
SQLite: xamarin/sqlite/3.31.1@49232bc
Xamarin.Android Tools: xamarin/xamarin-android-tools/d16-6@bfb66f3

=== Microsoft OpenJDK for Mobile ===

Java SDK: /Users/stefanglr/Library/Developer/Xamarin/jdk/microsoft_dist_openjdk_1.8.0.25
1.8.0-25
Android Designer EPL code available here:
https://github.com/xamarin/AndroidDesigner.EPL

=== Android SDK Manager ===

Version: 16.6.0.50
Hash: 5901879
Branch: remotes/origin/d16-6
Build date: 2020-06-10 22:42:50 UTC

=== Android Device Manager ===

Version: 16.6.0.96
Hash: 6e8b80b
Branch: remotes/origin/d16-6
Build date: 2020-06-10 22:43:28 UTC

=== Apple Developer Tools ===

Xcode 11.5 (16139)
Build 11E608c

=== Xamarin.Mac ===

Version: 6.18.2.1 (Visual Studio Community)
Hash: 29c4ea731
Branch: d16-6
Build date: 2020-05-26 17:03:04-0400

=== Xamarin.iOS ===

Version: 13.18.2.1 (Visual Studio Community)
Hash: 29c4ea731
Branch: d16-6
Build date: 2020-05-26 17:03:05-0400

=== Xamarin Designer ===

Version: 16.6.0.329
Hash: d4f8bcd13
Branch: remotes/origin/d16-6
Build date: 2020-04-24 02:16:02 UTC

=== Build Information ===

Release ID: 806050023
Git revision: 5289d413b99fddfc20b4ecf3e445ccb822213427
Build date: 2020-06-18 12:08:30-04
Build branch: release-8.6
Xamarin extensions: 5289d413b99fddfc20b4ecf3e445ccb822213427

=== Operating System ===

Mac OS X 10.15.5
Darwin 19.5.0 Darwin Kernel Version 19.5.0
    Thu Apr 30 18:25:59 PDT 2020
    root:xnu-6153.121.1~7/RELEASE_X86_64 x86_64

=== Enabled user installed extensions ===

VsVim 2.8.0.7
```

### PR DESCRIPTION
**Pull request by mandel-macaque**
This is an interesting issue that happens in the following scenario:

1. A request is performed that will take some extra time to get a
   response or an error.
2. The request is cancelled BEFORE the delegate methods
   DidReceiveResponse, DidReceiveData or DidCompleteWithError and
   called. Yet we have not yet fully canceled the native request. It
   does take some time to do so, that is why we have a
   'NSURLSessionTaskStateCanceling' and not 'NSURLSessionTaskStateCancelled'

The issue happens because the GetInflightData checks if the task has
been canceled. In that method we have the following:

```csharp
if (inflight.CancellationToken.IsCancellationRequested)
  task?.Cancel ();
return inflight;
```

The call of the Cancel method in the NSData task has as ripple effect which
is the execution of the following callback:

```csharp
cancellationToken.Register (() => {
  RemoveInflightData (dataTask);
  tcs.TrySetCanceled ();
});
```

Which calls:

```csharp
void RemoveInflightData (NSUrlSessionTask task, bool cancel = true)
{
      lock (inflightRequestsLock) {
              if (inflightRequests.TryGetValue (task, out var data)) {
                      if (cancel)
                              data.CancellationTokenSource.Cancel ();
                      data.Dispose ();
                      inflightRequests.Remove (task);
              }
              // do we need to be notified? If we have not inflightData, we do not
              if (inflightRequests.Count == 0)
                      RemoveNotification ();
      }

      if (cancel)
              task?.Cancel ();

      task?.Dispose ();
}
```

The above call does call Dispose and Dipose in the inflight data does:
```csharp
if (disposing) {
    if (CancellationTokenSource != null) {
            CancellationTokenSource.Dispose ();
            CancellationTokenSource = null;
    }
}
```

If we follow these set of events for example in the
DidRecieveResponse:
```csharp
[Preserve (Conditional = true)]
public override void DidReceiveResponse (NSUrlSession session, NSUrlSessionDataTask dataTask, NSUrlResponse response, Action<NSUrlSessionResponseDisposition> completionHandler)
{
    var inflight = GetInflightData (dataTask);

    if (inflight == null)
            return;

    try {
            var urlResponse = (NSHttpUrlResponse)response;
            var status = (int)urlResponse.StatusCode;

            var content = new NSUrlSessionDataTaskStreamContent (inflight.Stream, () => {
                    if (!inflight.Completed) {
                            dataTask.Cancel ();
                    }

                    inflight.Disposed = true;
                    inflight.Stream.TrySetException (new ObjectDisposedException ("The content stream was disposed."));

                    sessionHandler.RemoveInflightData (dataTask);
            }, inflight.CancellationTokenSource.Token);
```
If can happen that, when we get the delegate call to the method, the
request has been cancelled. That means that we are in the precise state
in which we do not longer care about the response BUT we did get a
infligh data reference.. that HAS BEEN DISPOSED!!! this later gives a NRE
in several places.

The correct way is to return null from the GetInflighData method if we
have been cancelled, this makes sense because if we cancelled we do not
care about the execution of any of the methods that required the data
since it has been disposed!

fixes #9132

### CODE REVIEW

File: src/Foundation/NSUrlSessionHandler.cs [hunk 1]
```diff
@@ -542,9 +542,20 @@ InflightData GetInflightData (NSUrlSessionTask task)
...
 				lock (sessionHandler.inflightRequestsLock)
 					if (sessionHandler.inflightRequests.TryGetValue (task, out inflight)) {
-						// ensure that we did not cancel the request, if we did, do cancel the task
-						if (inflight.CancellationToken.IsCancellationRequested)
+						// ensure that we did not cancel the request, if we did, do cancel the task, if we 
+						// cancel the task it means that we are not interested in any of the delegate methods:
+						// 
+						// DidReceiveResponse     We might have received a response, but either the user cancelled or a 
+						//                        timeout did, if that is the case, we do not care about the respose.
+						// DidReceiveData         Of buffer has a partial response ergo garbage and there is not real
```
Comment (spouliot) [inline src/Foundation/NSUrlSessionHandler.cs:550]: ergo ? we're now using latin for comments :)

File: src/Foundation/NSUrlSessionHandler.cs [hunk 2]
```diff
@@ -542,9 +542,20 @@ InflightData GetInflightData (NSUrlSessionTask task)
...
+						// 
+						// DidReceiveResponse     We might have received a response, but either the user cancelled or a 
+						//                        timeout did, if that is the case, we do not care about the respose.
+						// DidReceiveData         Of buffer has a partial response ergo garbage and there is not real 
+						//                        reason we would like to add more data.
+						// DidCompleteWithError - We are not changing a behaviour compared to the case in which 
+						//                        we did not find the data.
+						if (inflight.CancellationToken.IsCancellationRequested) {
 							task?.Cancel ();
+							// return null so that we break out of any deleate method.
```
Comment (spouliot) [inline src/Foundation/NSUrlSessionHandler.cs:556]: typo -> dele[g]ate

**Review by tj-devel709 — APPROVED** [review]

**Review by spouliot — APPROVED** [review]
Let's get confirmation it fix the customers issues before backporting to release branches (since it looks unlikely to get a [unit] test for this)

File: src/Foundation/NSUrlSessionHandler.cs (hunk 1 above)
Comment (mandel-macaque) [reply src/Foundation/NSUrlSessionHandler.cs:550]: We are that posh :)

**Comment by mandel-macaque** [comment]
@spouliot I'll trigger a pkg build for this.

**Review by dalexsoto — APPROVED** [review]

**Comment by mandel-macaque** [comment]
Watch issues are known and have been fixed on https://github.com/xamarin/xamarin-macios/pull/10235

File: src/Foundation/NSUrlSessionHandler.cs [hunk 3]
```diff
@@ -542,9 +542,20 @@ InflightData GetInflightData (NSUrlSessionTask task)
 
 				lock (sessionHandler.inflightRequestsLock)
 					if (sessionHandler.inflightRequests.TryGetValue (task, out inflight)) {
-						// ensure that we did not cancel the request, if we did, do cancel the task
-						if (inflight.CancellationToken.IsCancellationRequested)
+						// ensure that we did not cancel the request, if we did, do cancel the task, if we 
+						// cancel the task it means that we are not interested in any of the delegate methods:
+						// 
+						// DidReceiveResponse     We might have received a response, but either the user cancelled or a 
+						//                        timeout did, if that is the case, we do not care about the respose.
```
Comment (rolfbjarne) [inline src/Foundation/NSUrlSessionHandler.cs:549]:
```suggestion
						//                        timeout did, if that is the case, we do not care about the response.
```

**Review by rolfbjarne — APPROVED** [review]

**Comment by ahhassan3** [comment]
Is there any estimate on when will this fix be released ? It's causing us many crashes

**Comment by mandel-macaque** [comment]
@ahhassan3 we have not landed it because we did not had customers confirming the fix.

**Comment by ahhassan3** [comment]
@mandel-macaque This bug is somewhat difficult to reproduce on demand but we have ad over 70 counts of crashes from the customer who use our app because of this issue. I saw that you have merged it on to the main code branch, do you know when the xamarin.forms nuget will be updated with this fix in it. Thank you

**Pull request merged**